Python/pywarpx/diagnostics/data/BaseReader.py

```diff
@@ -0,0 +1,173 @@
+                          'path:\n  {}\n'
+                          'Did you set the proper path to the run directory?\n'
+                          'Did the simulation already run?'
+                          .format(self.run_directory))
+
+        data_file_path = os.path.join(
+            sim_output_dir,
+            self.data_file_prefix +
+            self.data_file_suffix
+        )
+        if not os.path.isfile(data_file_path):
+            raise IOError('The file {} does not exist.\n'
+                          'Did the simulation already run?'
+                          .format(data_file_path))
+
+        return data_file_path
+
+    def get_ncols(self):
+        ''' Returns the number of columns of the reduced diagnostic.'''
+        data_file_path = self.get_data_path()
+        data = np.loadtxt(data_file_path)
+        if data.ndim > 1:
+            ncols = np.shape(data)[1]
+        else:
+            ncols = len(data)
+        return ncols
+
+    def get_nspecies(self, subtract=None, divide=None):
+        '''Returns the number of species involved in the diagnostic.'''
+        # deduces number of species from number of columns
+        if self.has_species == True:
+            ncols = self.get_ncols()
+            # remove <substract> columns and divide to avoid counting multiple entries
+            nspecies = int((ncols - subtract)/divide)
+            return nspecies
+        else:
+            return 0
+
+    def get_species_names(self, string=None, start=None, step=None):
+        '''Returns the names of the species involved in the diagnostic.'''
+        # deduces names of species from the header by considering only the relevant entries
+        # finds the string by locating the units
+        if self.has_species == True:
+            species_names = []
+            data_file_path = self.get_data_path()
+            nspecies = self.get_nspecies()
+            with open(data_file_path) as f:
+                for line in f:
+                    if line.startswith('#'):
+                        data = line.split()
+                        for i in range(nspecies):
+                            result=re.search('](.+?)'+string,data[start+i*step])
+                            species_names.append(result.group(1))
+            return species_names
+        else:
+            raise ValueError('The current diagnostics does not involve any species!')
+
+    def get_steps(self):
+        data_file_path = self.get_data_path()
+        df = pd.read_csv(data_file_path, sep=" ", header=0)
+        steps = np.unique(df['#[0]step()'].values)
+        return steps
+
+    def get_times(self):
+        data_file_path = self.get_data_path()
+        df = pd.read_csv(data_file_path, sep=" ", header=0)
+        times = np.unique(df['[1]time(s)'].values)
+        return times
+
+
+    def restrict_data(self, data, **kwargs): # check if works for all diags (e.g. probes, field reduction)
```

Comment:
This would be cleaner python if the keyword arguments were specified
`def restrict_data(self, data, steps=None, times=None):`
This would make the logic simpler, something like below:
```
if steps is not None and times is not None:
    raise Exception()
elif steps is not None:
    if not isinstance(steps, collections.abc.Iterable):
         steps = np.array([steps])
elif times is not None:
    # set steps from times
if steps is None:
    steps = self.get_steps()
```

Reply:
Agreed.